docs/design-proposals/2023-11 TraceQL Metrics.md

```diff
@@ -0,0 +1,112 @@
+|`avg_over_time(<field>)`                    | The average of a numeric value like `duration` or [http.request.body.size](https://opentelemetry.io/docs/specs/semconv/attributes-registry/http/)
+|`max_over_time(<field>)`                    | The maximum of a numeric value.
+|`min_over_time(<field>)`                    | The minimum of a numeric value.
+|`quantile_over_time(<field>, q1, q2, ...)`  | The quantile of a numeric value (e.g. p95). Multiple quantiles can be requested and a time-series is created for each. |
+
+### Interval
+All aggregations take an optional interval like `30s`, `5m`, `1h`. If not specified then the interval will automatically match the step interval of the query. The interval is the last parameter of each function.
+
+`rate(5m)`  
+`quantile_over_time(duration, 0.95, 5m)`
+
+### Grouping
+All aggregations can group by one or more attributes and generate a time-series for each combination of values.  
+
+Plot the request rate of `/myapi` by user ID and HTTP status code:  
+`{ span.http.path = "/myapi" } | rate() by (span.user_id, span.http.status_code)`
+
+Plot the p95 duration of requests to service by HTTP endpoint:  
+`{ resource.service.name = "myservice" } | quantile_over_time(duration, 0.95) by (span.http.path)`
+
+### Filtering
+Aggregations can be combined with comparison operations to filter the output.
+
+Find the request rate per service but only if more than 1000 req/s:  
+`{ } | rate() by (resource.service.name) > 1000`
+
+
+## Additional Stages
+Additional metrics stages can be added to further refine and transform the time-series. These stages receive time-series and generate new time-series so they are separate from the first stage.  
+
+Most functions support the same ability to regroup using `by()` and comparison operators to refilter.  These can be thought of as instant functions that work across inputs at each point in time. Therefore they do not have an interval and do not alter the frequency of data points. If the original aggregation creates a data point every 60s, these will also output a data point every 60s.
+
+| function                        | description                |
+----------------------------------| ---------------------------- 
+| `... \| max() [by(...)]`        | Maximum value for each point in time
+| `... \| min() [by(...)]`        | Minimum value for each point in time
+| `... \| avg() [by(...)]`        | Average across values for each point in time
+| `... \| stddev() [by(...)]`     | Standard deviation across values for each point in time
+| `... \| quantile(q) [by(...)]`  | Quantile across values for each point in time
+| `... \| topk(N)`                | Return the top series for each point in time (grouping TBD)
+
+Find the highest per-pod failure rate in each cluster:  
+`{ status = error } | rate() by (cluster, pod) | max() by (cluster)`
+
+
+## Arithmetic
+Arithmetic operations between time-series can be used to calculate things like error rates.
```

Comment:
if we're defining operators between time series, should we also discuss join semantics?

Reply:
Agree, @cyriltovena 's [comment](https://github.com/grafana/tempo/pull/3138#discussion_r1391738304) about comparison also highlighted this gap.   Maybe we don't need the full featureset of PromQL [vector matching](https://prometheus.io/docs/prometheus/latest/querying/operators/#vector-matching).

Reply:
Added this to the Future section.  It's TBD, we can define it once we add the features that need it like time-series arithmetic.